docs/cs-proof.md

```diff
@@ -0,0 +1,800 @@
+
+Then, the prover performs a combination of the following operations to generate the constraints:
+
+1. **Allocate a multiplier:** a new [multiplication gate](#multiplication-gates) is added, represented by three [low-level variables](#variables) \\(a_L, a_R, a_O\\), for left input, right input and output value respectively.
+2. **Add a constraint:** a [linear combination](#linear-constraints) of any number of [variables](#variables) is encoded into appropriate positions in matrices \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O, \mathbf{W}\_V\\) and a vector of constants \\(\mathbf{c}\\).
+3. **Request a challenge scalar:** a random scalar returned in response to committed [high-level variables](#variables).
+
+
+### Uncommitted variables
+
+Often a constraint system is composed of _gadgets_: subsets of constraints
+that implement features of a higher-level protocol.
+The “wires” that connect the gadgets are called _uncommitted variables_.
+
+Uncommitted variables are created from left and right variables \\(a\_L, a\_R\\) of additional multiplication gates.
+Output variables \\(a\_O\\) are not used for this purpose because
+they are implicitly constrained by a [multiplication gate](#multiplication-gates)
+and cannot be used as independent uncommitted variables.
+
+
+### Gadget as a challenge
+
+Intermediate challenge scalars can be used to construct gadgets more efficiently.
+
+For example, a proof of permutation (“verifiable shuffle”) can be done by proving equality of
+two polynomials sampled at a challenge point, where roots of each polynomial
+represent secret values of the corresponding side of a permutation:
+
+\\[
+ \\{a,b\\} = \\{c,d\\}  \iff  (a-x)\cdot(b-x) = (c-x)\cdot(d-x),
+\\] where \\(x\\) is a random challenge, sampled after all values \\(a,b,c,d\\) are committed.
+
+Making a proof of permutation using a static gadget (without challenge values) may require
+building a [sorting network][sorting_network] that would use significantly more multiplication gates.
+
+**Important:** since challenges in the linear constraints are bound only to [high-level variables](#variables),
+using them to implement high-level protocols requires proving the soundness of the system as a _whole_.
+Individual gadgets can be proven to be sound only when they do not use such challenges.
+This caveat does not apply to the challenges used to implement the underlying proof system (\\(y, z, x, w\\)),
+as these are produced after all variables (including low-level ones) are committed.
+
+[sorting_network]: https://en.wikipedia.org/wiki/Sorting_network
+
+
+### Representation of constraints
+
+The matrices \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O, \mathbf{W}\_V\\) are typically very sparse
+because most constraints apply only to a few variables. As a result, constraints are represented as short lists
+of pairs \\((i, w)\\) where \\(i\\) is an index of a variable, and `w` is its (non-zero) weight.
```

Comment:
the formatting for `w` is different than the other math formatting \\(i\\)

Reply:
👍